src/types/number.rs

```diff
@@ -0,0 +1,210 @@
+        col.clone().slice(range.start, range.end - range.start)
+    }
+
+    fn iter_column<'a>(col: &'a Self::Column) -> Self::ColumnIterator<'a> {
+        col.iter().cloned()
+    }
+
+    fn column_from_iter(iter: impl Iterator<Item = Self::Scalar>, _: &GenericMap) -> Self::Column {
+        iter.collect()
+    }
+
+    fn create_builder(capacity: usize, _generics: &GenericMap) -> Self::ColumnBuilder {
+        Vec::with_capacity(capacity)
+    }
+
+    fn column_to_builder(col: Self::Column) -> Self::ColumnBuilder {
+        buffer_into_mut(col)
+    }
+
+    fn builder_len(builder: &Self::ColumnBuilder) -> usize {
+        builder.len()
+    }
+
+    fn push_item(builder: &mut Self::ColumnBuilder, item: Self::Scalar) {
+        builder.push(item);
+    }
+
+    fn push_default(builder: &mut Self::ColumnBuilder) {
+        builder.push(T::Storage::default());
+    }
+
+    fn append_builder(builder: &mut Self::ColumnBuilder, other_builder: &Self::ColumnBuilder) {
+        builder.extend_from_slice(other_builder);
+    }
+
+    fn build_column(builder: Self::ColumnBuilder) -> Self::Column {
+        builder.into()
+    }
+
+    fn build_scalar(builder: Self::ColumnBuilder) -> Self::Scalar {
+        assert_eq!(builder.len(), 1);
+        builder[0]
+    }
+}
+
+impl Number for u8 {
+    type Storage = u8;
+
+    fn data_type() -> DataType {
+        DataType::UInt8
+    }
+
+    fn try_downcast_scalar(scalar: &Scalar) -> Option<Self::Storage> {
+        scalar.as_u_int8().cloned()
+    }
+
+    fn try_downcast_column(col: &Column) -> Option<Buffer<Self::Storage>> {
+        col.as_u_int8().cloned()
+    }
+
+    fn upcast_scalar(scalar: Self::Storage) -> Scalar {
+        Scalar::UInt8(scalar)
+    }
+
+    fn upcast_column(col: Buffer<Self::Storage>) -> Column {
+        Column::UInt8(col)
+    }
+}
+
+impl Number for u16 {
+    type Storage = u16;
+
+    fn data_type() -> DataType {
+        DataType::UInt16
+    }
+
+    fn try_downcast_scalar(scalar: &Scalar) -> Option<Self::Storage> {
+        scalar.as_u_int16().cloned()
+    }
+
+    fn try_downcast_column(col: &Column) -> Option<Buffer<Self::Storage>> {
+        col.as_u_int16().cloned()
+    }
+
+    fn upcast_scalar(scalar: Self::Storage) -> Scalar {
+        Scalar::UInt16(scalar)
+    }
+
+    fn upcast_column(col: Buffer<Self::Storage>) -> Column {
+        Column::UInt16(col)
+    }
+}
+
+impl Number for i8 {
+    type Storage = i8;
+
+    fn data_type() -> DataType {
+        DataType::Int8
+    }
+
+    fn try_downcast_scalar(scalar: &Scalar) -> Option<Self::Storage> {
+        scalar.as_int8().cloned()
+    }
+
+    fn try_downcast_column(col: &Column) -> Option<Buffer<Self::Storage>> {
+        col.as_int8().cloned()
+    }
+
+    fn upcast_scalar(scalar: Self::Storage) -> Scalar {
+        Scalar::Int8(scalar)
+    }
+
+    fn upcast_column(col: Buffer<Self::Storage>) -> Column {
+        Column::Int8(col)
+    }
+}
+
+impl Number for i16 {
```

Comment:
using a macro to implement could be better.

Reply:
We can also implement Timestamp as a Number type.